server/global/data_analyzer/analyzer.js

```diff
@@ -0,0 +1,298 @@
+ * Returns values that appear more than once in the list
+ * @param {any[]} values
+ * @returns {any[]} non-unique values
+ */
+function getDuplicates(values) {
+  return [...new Set(values.filter((value, i) => values.slice(i + 1).includes(value)))];
+}
+
+/**
+ * Get all values which are longer that the max length
+ * @param {string[]} values
+ * @param {number} maxlength
+ * @returns {string[]} The list a too long values
+ */
+function getTooLongValues(values, maxlength) {
+  return values.filter((value) => value.length > maxlength);
+}
+
+/**
+ * Returns all groups of values that have a distance less than a given one
+ * @param {string[]} values The values
+ * @param {number} minDistance The maximum distance
+ * @returns {string[][]}
+ */
+function getTooCloseValues(values, minDistance) {
+  values = values.slice();
+  const groups = [];
+
+  values.forEach((value) => {
+    if (!isString(value)) {
+      return;
+    }
+    const group = values.filter((other) => {
+      if (!isString(other)) {
+        return;
+      }
+      const distance = levenshtein(other, value);
+      return distance <= minDistance && distance > 0;
+    });
+
+    // console.log(value, group);
+
+    const existingGroup = groups.find((egroup) => egroup.some((e) => group.includes(e)));
+    if (existingGroup) {
+      existingGroup.push(...group, value);
+      return;
+    }
+
+    if (group.length > 0) {
+      groups.push([...group, value]);
+    }
+  });
+
+  return groups.map((group) => [...new Set(group)]);
+}
+
+/**
+ * Find nodes with the same name but a different parent
+ * @param {object[]} classification The set of nodes
+ * @returns {{node : string, parents : string[]}[]}
+ */
+function findNodeWithDifferentsParents(classification) {
+  classification = flattenClassification(classification);
+  const groups = [];
+  classification.forEach((node, i) => {
+    if (!node) {
+      return;
+    }
+
+    const duplicates = classification
+      .slice(i)
+      .filter((other) => other && node.name === other.name)
+      .map((node) => node.name);
+
+    if (duplicates.length > 1) {
+      const parents = classification
+        .filter((n) => n && n.children.find((n) => n.name === duplicates[0]))
+        .map((p) => p.name);
+
+      groups.push({ node: duplicates[0], parents });
+      classification = classification.map((n) => (duplicates.includes(n) ? null : n));
+    }
+  });
+  return groups;
+}
+
+/**
+ * Return an array of all node of all level of a classification
+ * @param {ClassificationNode[]} classification
+ * @returns {ClassificationNode[]}
+ */
+function flattenClassification(classification) {
+  function flattenNode(node) {
+    let res = [node];
+    for (let child of node.children) {
+      res.push(...flattenNode(child));
+    }
+    return res;
+  }
+
+  const res = [];
+  for (let node of classification) {
+    res.push(...flattenNode(node));
+  }
+  return res;
+}
+
+const isString = (v) => typeof v === "string" || v instanceof String;
```

Comment:
In a real developer's language, you could have created a template function like `isType<T>(v)`! But hey, you preferred to do JS xD

Reply:
😅